internal/output_sse.h

```diff
@@ -0,0 +1,304 @@
+// it all at once, instead of as 4 separate int32x4 values as in the above
+// generic partial specialization. This also avoids the poor (50%) register
+// utilization of FragmentUint8x4x1: by handling 16 scalar values at once,
+// we are able to fill a uint8x16_t.
+template <>
+struct OutputStageEvalImpl<OutputStageSaturatingCastToUint8,
+                           SSE4FragmentInt32x16x1> {
+  typedef SSE4FragmentInt32x16x1 InputType;
+  typedef SSE4FragmentUint8x16x1 OutputType;
+  typedef OutputStageSaturatingCastToUint8 OutputStage;
+
+  OutputStageEvalImpl(const OutputStage&) {}
+
+  OutputType Eval(InputType input, int, int) const {
+    __m128i q16[2];
+    for (int i = 0; i < 2; i++) {
+      q16[i] = _mm_packus_epi32(input.data.val[2 * i],
+				input.data.val[2 * i + 1]);
+    }
+    return _mm_packus_epi16(q16[0], q16[1]);
+  }
+};
+
+// Implementation of OutputStageBiasAddition for SSE4FragmentInt32x4x1
+template <typename VectorType>
+struct OutputStageEvalImpl<OutputStageBiasAddition<VectorType>,
+                           SSE4FragmentInt32x4x1> {
+  typedef SSE4FragmentInt32x4x1 InputType;
+  typedef SSE4FragmentInt32x4x1 OutputType;
+  typedef OutputStageBiasAddition<VectorType> OutputStage;
+
+  OutputStageEvalImpl(const OutputStage& s) : output_stage(s) {}
+
+  OutputType Eval(InputType input, int row, int col) const {
+    __m128i bias;
+    if (VectorType::kShape == VectorShape::Row) {
+      bias = _mm_set1_epi32(output_stage.bias_vector(col));
+    } else {
+      bias = _mm_lddqu_si128(reinterpret_cast<__m128i*>
+			     (output_stage.bias_vector.data(row)));
+    }
+    return _mm_add_epi32(input, bias);
+  }
+
+  const OutputStage& output_stage;
+};
+
+// Implementation of OutputStageClamp for SSE4FragmentInt32x4x1
+template <>
+struct OutputStageEvalImpl<OutputStageClamp, SSE4FragmentInt32x4x1> {
+  typedef SSE4FragmentInt32x4x1 InputType;
+  typedef SSE4FragmentInt32x4x1 OutputType;
+  typedef OutputStageClamp OutputStage;
+
+  OutputStageEvalImpl(const OutputStage& s) : output_stage(s) {}
+
+  OutputType Eval(InputType input, int, int) const {
+    const __m128i min = _mm_set1_epi32(output_stage.min);
+    const __m128i max = _mm_set1_epi32(output_stage.max);
+    return _mm_min_epi32(_mm_max_epi32(input, min), max);
+  }
+
+  const OutputStage& output_stage;
+};
+
+// Implementation of OutputStageTanh for SSE4FragmentInt32x4x1
+template <>
+struct OutputStageEvalImpl<OutputStageTanh, SSE4FragmentInt32x4x1>
+    : OutputStageTanhEvalImpl<SSE4FragmentInt32x4x1> {
+  OutputStageEvalImpl(const OutputStageTanh& output_stage)
+      : OutputStageTanhEvalImpl(output_stage) {}
+};
+
+// Specialization of StoreFinalOutput for SSE4FragmentUint8x4x1.
+template <typename DstType>
+inline void StoreFinalOutput(SSE4FragmentUint8x4x1 value, DstType* dst, int row,
+                             int col) {
+  uint32_t *tmp = reinterpret_cast<uint32_t *>(dst->data(row,col));
```

Comment:
That is a correct concern.
According to the C11 standard we are not allowed to do so. I can replace this with 4 stores of 8bit and not cast the pointer. This will lead the compiler to issue 4 mov and 3 shift right instructions.
However, in practice, as x86 machines are byte addressable, the code above results only in a mov instruction (which does not care about alignment).
So I let you pick which one you think is the best.